realm/realm-library/src/objectServer/java/io/realm/PermissionManager.java

```diff
@@ -0,0 +1,517 @@
+            close();
+        }
+        super.finalize();
+    }
+
+    // Creates the URL to the permission/management Realm based on the authentication URL.
+    private static String getRealmUrl(RealmType type, URL authUrl) {
+        String scheme = "realm";
+        if (authUrl.getProtocol().equalsIgnoreCase("https")) {
+            scheme = "realms";
+        }
+        try {
+            return new URI(scheme, authUrl.getUserInfo(), authUrl.getHost(), authUrl.getPort(),
+                    "/~/" + type.getName(), null, null).toString();
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException("Could not create URL to the " + type + " + Realm", e);
+        }
+    }
+
+    // Task responsible for loading the Permissions result and returning it to the user.
+    // The Permission result is not considered available until the query has completed.
+    private class GetPermissionsAsyncTask extends PermissionManagerAsyncTask<RealmResults<Permission>> {
+
+        // Prevent permissions from being GC'ed until fully loaded.
+        private RealmResults<Permission> loadingPermissions;
+
+        GetPermissionsAsyncTask(PermissionManager permissionManager, Callback<RealmResults<Permission>> callback) {
+            super(permissionManager, callback);
+        }
+
+        @Override
+        public void run() {
+            if (checkAndReportInvalidState()) { return; }
+            if (permissions != null) {
+                // Permissions already loaded
+                notifyCallbackWithSuccess(permissions);
+            } else {
+                // Start loading permissions.
+                // TODO Right now multiple getPermission() calls will result in multiple
+                // queries being executed. The first one to return will the be one returned
+                // by all callbacks.
+                loadingPermissions = permissionRealm.where(Permission.class).findAllAsync();
+                loadingPermissions.addChangeListener(new RealmChangeListener <RealmResults<Permission>>() {
+                    @Override
+                    public void onChange(RealmResults <Permission> loadedPermissions) {
+                        loadingPermissions.removeChangeListener(this);
+                        if (checkAndReportInvalidState()) { return; }
+                        if (permissions == null) {
+                            permissions = loadedPermissions;
+                        }
+                        notifyCallbackWithSuccess(permissions);
+                    }
+                });
+            }
+        }
+    }
+
+    // Class encapsulating all async tasks exposed by the PermissionManager.
+    // All subclasses are responsible for removing themselves from the activeTaskList when done.
+    // Made package protected instead of private to facilitate testing
+    abstract static class PermissionManagerAsyncTask<T> implements RealmAsyncTask, Runnable {
```

Comment:
the name might be redundant (already calling it as an inner class of `PermissionManager`) `AsyncTask` IMO should be clear enough (`new PermissionManager.AsyncTask`)